### PR TITLE
#773 fixed by checking not null insted of null

### PR DESCRIPTION
**Pull request by heshanjse**
Signed-off-by: Heshan Jayasinghe <shanujse@gmail.com>

This PR addresses GitHub issue: # 773 .

Briefly describe the changes proposed in this PR:

* fixed 773 by checking not null instead of null in RDFJSONParser class void parse(Reader reader,
String baseURI) method
* added two test-cases to AbstractParserHandlingTest class to test that issue

### CODE REVIEW

**Comment by hmottestad** [comment]
@heshanjse, It seems like the test is failing for some of the other parsers. I think the first step is to change the input to try to force the other parsers to error out with RDFParseException. Then we can have a look at the left over ones to see if they should be failing or if their behaviour is correct. If the behaviour is correct, then we need to move the test so that it doesn't run for those parsers.

File: testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java [hunk 1]
```diff
@@ -948,4 +951,15 @@ private final Model getTestModel(String languageValue, String languageTag) {
 				vf.createLiteral(languageValue, languageTag)));
 		return result;
 	}
+
+	@Test(expected = RDFParseException.class)
```
Comment (ansell) [inline testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java]: There need to be other tests where a valid document stream is sent in and parsing succeeds even with the rdfHandler being set to a non-null value.

**Comment by ansell** [comment]
Travis is saying that the tests that are in this PR are not passing.

File: testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java (hunk 1 above)
Comment (hmottestad) [inline testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java]: Purpose of the test is to make parsing in presence of a null rdfHandler consistent. So I propose to give it input other than "" to force it to never be valid RDF across all parsers. Maybe something like "{(.;".

**Comment by jeenbroekstra** [comment]
This is the output from Travis:

```
Failed tests: 
  NQuadsHandlingTest.testNullRdfHandlerInputStream Expected exception: org.eclipse.rdf4j.rio.RDFParseException
  NQuadsHandlingTest.testNullRdfHandlerReader Expected exception: org.eclipse.rdf4j.rio.RDFParseException
  NTriplesHandlingTest.testNullRdfHandlerInputStream Expected exception: org.eclipse.rdf4j.rio.RDFParseException
  NTriplesHandlingTest.testNullRdfHandlerReader Expected exception: org.eclipse.rdf4j.rio.RDFParseException
  TriGHandlingTest.testNullRdfHandlerInputStream Expected exception: org.eclipse.rdf4j.rio.RDFParseException
  TriGHandlingTest.testNullRdfHandlerReader Expected exception: org.eclipse.rdf4j.rio.RDFParseException
  TurtleHandlingTest.testNullRdfHandlerInputStream Expected exception: org.eclipse.rdf4j.rio.RDFParseException
  TurtleHandlingTest.testNullRdfHandlerReader Expected exception: org.eclipse.rdf4j.rio.RDFParseException
Tests in error: 
  BinaryHandlingTest.testNullRdfHandlerReader Â»  Unexpected exception, expected<...
```

File: testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java [hunk 2]
```diff
@@ -948,4 +951,30 @@ private final Model getTestModel(String languageValue, String languageTag) {
 				vf.createLiteral(languageValue, languageTag)));
 		return result;
 	}
-}
+//
+//	@Test(expected = RDFParseException.class)
```
Comment (hmottestad) [inline testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java]: clean up

File: testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java [hunk 3]
```diff
@@ -948,4 +951,30 @@ private final Model getTestModel(String languageValue, String languageTag) {
...
+//		getParser().parse(new InputStreamReader(new ByteArrayInputStream("".getBytes())) , "");
+//
+//	}
+
+	@Test
+	public void testNullRdfHandlerReader(){
+		try {
+			 getParser().parse(new InputStreamReader(new ByteArrayInputStream("".getBytes())), "");
+		}catch (Exception e){
+			assertTrue("Did not find expected rdfHandler", (!(e instanceof NullPointerException)));
```
Comment (hmottestad) [inline testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java]: Change your message to be more useful.

File: testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java [hunk 4]
```diff
@@ -948,4 +951,30 @@ private final Model getTestModel(String languageValue, String languageTag) {
...
+//	public void testNullRdfHandlerReader() throws IOException {
+//
+//		getParser().parse(new InputStreamReader(new ByteArrayInputStream("".getBytes())) , "");
+//
+//	}
+
+	@Test
+	public void testNullRdfHandlerReader(){
+		try {
+			 getParser().parse(new InputStreamReader(new ByteArrayInputStream("".getBytes())), "");
```
Comment (hmottestad) [inline testsuites/rio/src/main/java/org/eclipse/rdf4j/rio/AbstractParserHandlingTest.java]: force the rdfhandler to be null. Maybe you can store getParser() in a variable and call something like .setHandler(null).

File: core/rio/rdfjson/src/main/java/org/eclipse/rdf4j/rio/rdfjson/RDFJSONParser.java [hunk 5]
```diff
@@ -154,7 +154,7 @@ protected void reportFatalError(String msg, JsonLocation location)
 	public void parse(final Reader reader, final String baseUri)
 		throws IOException, RDFParseException, RDFHandlerException
 	{
-		if (rdfHandler == null) {
+		if (rdfHandler != null) {
```
Comment (hmottestad) [inline core/rio/rdfjson/src/main/java/org/eclipse/rdf4j/rio/rdfjson/RDFJSONParser.java]: pull and merge master into this branch.

**Comment by jamesrdf** [comment]
I'd like to get this into 2.2.2. @heshanjse will you be able resolve the conflicts this week by pulling down the latest from the master branch?

**Comment by ansell** [comment]
The issue isn't just conflicting files, as #773 has already been fixed:

https://github.com/eclipse/rdf4j/blob/master/core/rio/rdfjson/src/main/java/org/eclipse/rdf4j/rio/rdfjson/RDFJSONParser.java#L77

The core of this pull request would be implementing some tests that directly test whether all parsers can cope with handlers being null, as permitted by the API contract. The current tests do not do that directly.

In a future major version I would like the API to change to guarantee that getRDFHandler never returns null, and instead returns a stub RDFHandler if the user has not set one, which would avoid these issues in general and cleanup the parser code a little.

**Comment by heshanjse** [comment]
@ansell I created these test for current situation. with your future implmentation i think this test can be not correct. 

@jamesrdf I can resolve this with in this week but with the future implmentation i think this test can be useless.I haven't more experience in rdf4j so please tell me most suitable way.

**Comment by ansell** [comment]
Hi @heshanjse, nothing will be changing until the next major version (3.x) at least, so there needs to be a test that verifies null handlers are supported in the meantime.
